addons/cookoff/functions/fnc_cookOffBoxLocal.sqf

```diff
@@ -0,0 +1,83 @@
+#include "..\script_component.hpp"
+/*
+ * Author: KoffeinFlummi, commy2, kymckay, johnb43
+ * Start a cook-off in the given ammo box.
+ *
+ * Arguments:
+ * 0: Ammo box <OBJECT>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * cursorObject call ace_cookoff_fnc_cookOffBox
+ *
+ * Public: No
+ */
+
+params ["_box", "_killer", "_instigator", "_startTime", "_igniteTime", "_smokeTime"];
+
+// Make sure effects are cleaned up if box is deleted
+[QGVAR(addCleanupHandlers), _box] call CBA_fnc_localEvent;
+
+// These time checks are for JIP players
+private _delay = _startTime - CBA_missionTime + _igniteTime;
+
+if (_delay >= 0) then {
+    [{
+        params ["_box"];
+
+        private _effects = [];
+
+        // Box will start smoking
+        if (hasInterface) then {
+            private _smoke = "#particlesource" createVehicleLocal [0, 0, 0];
+            _smoke setParticleClass "AmmoSmokeParticles2";
+            _smoke attachTo [_box, [0, 0, 0]];
+
+            _effects pushBack _smoke;
+        };
+
+        if (isServer) then {
+            private _sound = createSoundSource ["Sound_Fire", ASLToAGL getPosASL _box, [], 0];
+            _smoke attachTo [_sound];
+
+            _effects pushBack _sound;
+        };
+
+        _box setVariable [QGVAR(effects), _effects];
+    }, _box, _delay] call CBA_fnc_waitAndExecute;
+};
+
+// Smoke happens later
+_delay = _delay + _smokeTime;
+
+if (_delay >= 0) then {
+    // Light the fire (also handles lighting)
+    [{
+        params ["_box", "_killer", "_instigator"];
+
+        if (hasInterface) then {
+            private _fire = "#particlesource" createVehicleLocal [0, 0, 0];
+
+            _fire setParticleClass "AmmoBulletCore";
+            _fire attachTo [_box, [0, 0, 0]];
+
+            private _effects = _box getVariable [QGVAR(effects), []];
+
+            _effects pushBack _fire;
+
+            _box setVariable [QGVAR(effects), _effects];
+        };
+
+        // These functions are smart and do all the cooking off work
+        if (isServer) then {
+            (_box call FUNC(getVehicleAmmo)) params ["_magazines", "_total"];
+
+            [QGVAR(detonateAmmunition), [_box, _magazines, _total, true, _killer, _instigator]] call CBA_fnc_localEvent;
+
+            // This shit is busy being on fire, magazines aren't accessible/usable
+            clearMagazineCargoGlobal _box;
```

Comment:
**Idea:** Not every mag, and individual cartridge (or the like) in a mag, will be gone at the same time, so it would be better if the contained mags would be iterated and given a random time to get destroyed, which would allow to save at least some ammo if you're quick enough, which makes it a lot more forgiving and realistic.

Reply:
Yes, but as it is, `FUNC(detonateAmmunition)` handles ammunition detonation entirely, disregarding the actual magazines in the vehicle/box.
It would be more realistic, but it would also be more performance intensive, as well as require some non-trivial redesign, meaning it's not guaranteed to happen.

Reply:
A shame, but fair enough.

Reply:
It's not going to happen, until BI fixes https://community.bistudio.com/wiki/magazineTurretAmmo and https://community.bistudio.com/wiki/setMagazineTurretAmmo. I rewrote what was necessary and added an extra function (see spoiler), but imo it's too intensive performance wise.

<details>
<summary>Function</summary>

```sqf
#include "..\script_component.hpp"
/*
 * Author: johnb43
 * Removes a magazine from the vehicle that has a specific ammo count.
 *
 * Arguments:
 * 0: Vehicle <OBJECT>
 * 1: Magazine <STRING>
 * 2: Turret path <ARRAY>
 * 3: Ammo count <NUMBER>
 *
 * Return Value:
 * Magazine removed <BOOL>
 *
 * Example:
 * [vehicle player, "12Rnd_120mm_HEAT_MP_T_Red", 12, [0]] call ace_common_fnc_removeSpecificMagazineTurret
 *
 * Public: No
 */

params [["_vehicle", objNull, [objNull]], ["_magazineType", "", [""]], ["_turretPath", [], [[]]], ["_ammoCount", 0, [0]]];

if !(_vehicle turretLocal _turretPath) exitWith {};

// Check how much ammo a weapon has
private _magazinesAmmo = [];
private _found = false;

{
    _x params ["_xMag", "_xTurret", "_xAmmo"];

    if ((_xMag == _magazineType) && {_xTurret isEqualTo _turretPath}) then {
        _magazinesAmmo pushBack _xAmmo;
    };
} forEach (magazinesAllTurrets _vehicle);

private _index = _magazinesAmmo find _ammoCount;

// If not found, exit
if (_index == -1) exitWith {false};

private _weapon = "";

// Find weapon that has mag (assumption: 1 magazine can only fit in 1 weapon per turret)
{
    if (_magazineType in (compatibleMagazines _x)) exitWith {
        _weapon = _x;
    };
} forEach (_vehicle weaponsTurret _turretPath);

if (_weapon == "") exitWith {false};

// Get current weapon
(weaponState [_vehicle, _turretPath]) params ["_currentWeapon", "", "", "_currentMagazine", "_currentAmmo"];

_vehicle removeMagazinesTurret [_magazineType, _turretPath];
_magazinesAmmo deleteAt _index;

// If the current magazine is not the specified magazine, there is nothing to worry about, so delete as such
if (_currentMagazine != _magazineType) exitWith {
    // Add old magazines back
    {
        _vehicle addMagazineTurret [_magazineType, _turretPath, _x];
    } forEach _magazinesAmmo;

    true
};

// Remove weapon
_vehicle removeWeaponTurret [_weapon, _turretPath];

// If the magazine that's to be removed is the current magazine, take extra steps
if (!isNil {_magazinesAmmo deleteAt (_magazinesAmmo find _currentAmmo)}) then {
    // Add old magazine back, then add weapon back, so that mag is instantly loaded (otherwise engine takes highest ammo magazine)
    _vehicle addMagazineTurret [_magazineType, _turretPath, _currentAmmo];
    _vehicle addWeaponTurret [_weapon, _turretPath];

    // Add rest of old magazines back
    {
        _vehicle addMagazineTurret [_magazineType, _turretPath, _x];
    } forEach _magazinesAmmo;
} else {
    // Add rest of old magazines back
    {
        _vehicle addMagazineTurret [_magazineType, _turretPath, _x];
    } forEach _magazinesAmmo;

    _vehicle addWeaponTurret [_weapon, _turretPath];
};

// Select old weapon again
_vehicle selectWeaponTurret [_currentWeapon, _turretPath];

true

```
</details>